core/src/multi_proof_verification.rs

```diff
@@ -0,0 +1,636 @@
+fn are_siblings(key_path1: &BitVec<u8, Msb0>, key_path2: &BitVec<u8, Msb0>) -> bool {
+    if key_path1.len() != key_path2.len() {
+        return false;
+    }
+
+    let d = key_path1.len() - 1;
+    key_path1[..d] == key_path2[..d]
+}
+
+// Accept two SubTrieRoots representing the left (a) and right (b) child nodes
+// that will be used to create a new internal node
+fn join_sub_trees<H: NodeHasher>(mut a: SubTrieRoot, b: SubTrieRoot) -> SubTrieRoot {
+    let parent = InternalData {
+        left: a.root.clone(),
+        right: b.root.clone(),
+    };
+    // remove last bit
+    let _ = a.key_path.pop();
+    SubTrieRoot {
+        root: H::hash_internal(&parent),
+        key_path: a.key_path,
+    }
+}
+
+// This helper function hashes up the tree starting from `curr_node` by `n_nodes`,
+// using the stack of previously formed sub_trees or an external sibling.
+// Newly created nodes are saved in `nodes_map`.
+fn hash_up_n_nodes<H: NodeHasher>(
+    curr_node: &mut SubTrieRoot,
+    n_nodes: usize,
+    sub_trie_stack: &mut Vec<SubTrieRoot>,
+    ext_siblings: &mut impl Iterator<Item = Node>,
+    nodes_map: &mut BTreeMap<BitVec<u8, Msb0>, Node>,
+) -> Result<(), MultiProofVerificationError> {
+    for bit in curr_node
+        .key_path
+        .clone()
+        .iter()
+        .by_vals()
+        .rev()
+        .take(n_nodes)
+    {
+        let sibling = match sub_trie_stack.last() {
+            Some(sub_tree_in_stack)
+                if are_siblings(&curr_node.key_path, &sub_tree_in_stack.key_path) =>
+            {
+                sub_trie_stack
+                    .pop()
+                    .expect("The existence of the last element in the stack has just been verified")
+                    .root
+            }
+            Some(_) | None => {
+                let ext_sibiling = ext_siblings
+                    .next()
+                    .ok_or(MultiProofVerificationError::IncorrectMultiProofFormat)?;
+
+                let mut sibling_path = curr_node.key_path.clone();
+                {
+                    let mut last_bit = sibling_path
+                        .last_mut()
+                        .expect("There must be at least one bit in the key_path");
+                    *last_bit = !last_bit.clone();
+                }
+
+                nodes_map.insert(sibling_path, ext_sibiling);
+                ext_sibiling
+            }
+        };
+
+        let parent = InternalData {
+            left: if bit { sibling } else { curr_node.root },
+            right: if bit { curr_node.root } else { sibling },
+        };
+
+        curr_node.root = H::hash_internal(&parent);
+        curr_node.key_path.pop();
+
+        nodes_map.insert(curr_node.key_path.clone(), curr_node.root);
+    }
+    Ok(())
+}
+
+impl<H: NodeHasher> VerifiableMultiProof<H> {
+    pub fn verify(
+        &self,
+        key_path: KeyPath,
+    ) -> Result<VerifiedPathProof, PathProofVerificationError> {
```

Comment:
I find this returning a `VerifiedPathProof` strange.

In my mind, the API of multi-proofs should look like this:

```rust
// build
fn multi_proof::create(Vec<PathProof>) -> MultiProof;
// or if we build incrementally...
fn multi_proof::MultiProofBuilder::build() -> MultiProof;

// verify or fail
fn multi_proof::verify(MultiProof) -> Result<VerifiedMultiProof, MultiProofVerificationError>;

struct VerifiedMultiProof {
    terminals: Vec<PathProofTerminal> // sorted by key path
}

fn VerifiedMultiProof::confirm_value(key_path, value_hash, index_hint: Option<usize>); // when hint not provided, do binary search.
fn VerifiedMultiProof::confirm_nonexistence(key_path, index_hint: Option<usize>); // when hint not provided, do binary search.
```

Reply:
I would expect that someone creating a proof would provide the index hints to the verifier, because they don't impact the validity of the proof, just the time complexity in getting information out of it.